app-drawer/app-drawer.html

```diff
@@ -260,70 +260,139 @@
-                  this._translateDrawer(translateDelta);
-                  this.$.scrim.style.opacity = 1 + translateDelta / this.getWidth();
-                  break;
-                case 'right':
-                  translateDelta = Math.max(0, translateDelta);
-                  this._translateDrawer(translateDelta);
-                  this.$.scrim.style.opacity = 1 - translateDelta / this.getWidth();
-                  break;
-              }
-
-              // Keep track of the last user track event with non-zero ddx (which we
-              // need to determine whether to keep the drawer open/closed) since the
-              // very last track event may have zero ddx.
-              if (event.detail.ddx !== 0) {
-                this._lastTrackDirection = event.detail.ddx;
-              }
+              this._track(event);
               break;
-
             case 'end':
-              this.$.contentContainer.style.transitionDuration = '';
-              this.$.scrim.style.transitionDuration = '';
-              this.$.scrim.style.visibility = '';
-              this.$.scrim.style.opacity = '';
-              this.transform('', this.$.contentContainer);
-
-              // Only toggle the drawer if the user intentionally swipes in one direction.
-              if (Math.abs(event.detail.dx) > 30) {
-                this.opened = (this.position === 'left' && this._lastTrackDirection > 0) ||
-                  (this.position === 'right' && this._lastTrackDirection < 0);
-              }
+              this._trackEnd(event);
               break;
           }
         }
       },
 
+      _trackStart: function(event) {
+        // Disable transitions since style attributes will reflect user track events.
+        this.$.contentContainer.style.transitionDuration = '0s';
+        this.$.scrim.style.transitionDuration = '0s';
+        this.$.scrim.style.visibility = 'visible';
+
+        if (this.opened) {
+          this._translateOffset = 0;
+        } else if (this.position == 'left') {
+          this._translateOffset = -this.getWidth();
+        } else if (this.position == 'right') {
+          this._translateOffset = this.getWidth();
+        }
+
+        this._trackEvents = [];
+      },
+
+      _track: function(event) {
+        this._translateDrawer(event.detail.dx + this._translateOffset);
+        this._trackEvents.push(event);
+      },
+
+      _trackEnd: function(event) {
+        // Save translate progress made during track events for this._rafAnimate.
+        this._translateOffset = event.detail.dx + this._translateOffset;
+        this._startTimeStamp = performance.now();
+
+        // Track handler takes precedence over scrim tap handler.
+        this.cancelDebouncer('_scrimTapHandler');
+
+        // Calculate the velocity with the first track event that is recent enough, if
+        // it exists
+        this._velocity = 0;
+        var timeLowerBound = Date.now() - 100;
```

Comment:
I see. My suggestion was that given that we might not need all the trackEvents, potentially we could just store an approximate track event unless we try to construct a curve fitting https://en.wikipedia.org/wiki/Curve_fitting. 

e.g. In case you only want to use a data point, I believe this might work: 

```js
trackStart: function(event) {
  this._lastEvent = event;
  this._secondLastEvent = event;
},
_track: function(event) {
  if (this._lastEvent.timeStamp + 100 < event.timeStamp) {
     this._secondLastEvent = this._lastEvent;
     this._lastEvent = event;
 }
},
trackEnd: function(event) {
  this._track(event);
  var dx = this._lastEvent.detail.dx -  this._secondLastEvent.detail.dx;
  var dt = this._lastEvent.timeStamp -  this._secondLastEvent.timeStamp;
  this._velocity = dx / dt;
}
```